日報機能追加、コメント機能の追加

ItohYuto opened the pull request:
- 日報を更新・削除できるのはその日報の作成者のみ
- 本と日報の各詳細画面からコメント投稿可能
- コメントを削除できるのはそのコメントの作成者のみ
- 投稿したコメントは本と日報の各詳細画面から確認できる。投稿した内容に加えて、投稿したユーザーの名前（未入力ならメアド）と投稿日時も表示
- コメントが一件もない場合は`コメントがありません`と表示する

ItohYuto committed:
日報機能追加、コメント機能の追加

ItohYuto committed:
日報に作成者と投稿日の表示を追加
コメントのフォームにrequired属性を追加

yuuu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -0,0 +1,22 @@
+    book = Book.find(params[:book_id])
+    comment = book.comments.build(comment_params)
+    comment.user = current_user
+    respond_to do |format|
```
今回の仕様であればJSONをレスポンスする必要はないのでこの `respond_to` は削除しておきましょう

ItohYuto replied:
修正しました。

yuuu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class Books::CommentsController < ApplicationController
+  def create
```
`Reports::CommentsController` の同じ名前のメソッドと処理の流れが同じになっていますね。メソッドを共通化できないか検討をお願い致します 🙏

ItohYuto replied:
`Commentable`モジュールを作成し処理を共通化しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,74 @@
+
+  # GET /reports/1/edit
+  def edit
+    redirect_to @report unless current_user.reports.exists?(params[:id])
```
この書き方でもOKですが、おそらくSQLクエリが余分に発行されるので以下のほうがシンプルですね。

```ruby
redirect_to @report if @report.user != current_user
```

ItohYuto replied:
おっしゃる通りですね。
修正しました。

yuuu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -0,0 +1,22 @@
+      if comment.save
+        format.html { redirect_to book_url(book), notice: t('controllers.common.notice_create', name: Book.model_name.human) }
+      else
+        format.html { render :new, status: :unprocessable_entity }
```
newに該当するViewがないので、コメントのバリデーションエラーが発生した場合にエラーが発生すると予想していますがいかがでしょうか？もしエラーが発生する場合は修正をお願い致します。

ItohYuto replied:
戻り先を追加対象の本の詳細ページに変更しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,74 @@
+  def create
+    @report = current_user.reports.build(report_params)
+
+    respond_to do |format|
```
このコントローラに関してもJSONを返す必要はないので `respond_to` は不要です。Viewの `json.jbuilder` ファイルも削除してOKです。

ItohYuto replied:
修正しました。

ItohYuto committed:
指摘箇所修正

- コメント追加時の処理を共通化
- JSONをレスポンス処理を削除。
- 日報編集ページへのアクセス時に余分なクエリが発生するため処理を変更。
- コメント追加時のエラー時の戻り先を修正。

ItohYuto commented:
修正しましたので、ご確認お願いいたします。

## スクリーンショット
- 日報一覧 
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/72d154df-d6e4-4b43-8285-25410c3670fc)
- 日報詳細ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/20efcf7b-38dd-4cc7-841e-aed2f5e60d43)
- 日報編集ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/6274235e-b9a8-4d3b-b87b-d1c8bcf7861a)
- 本詳細ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/ddcdf981-fa12-4da4-989c-f1097228e307)
- rubbocop実行結果
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/bca7e8aa-c5f6-4721-ab74-944b6c10f998)
- erblint実行結果
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/c41475e2-8178-4434-8408-c1c1e6bebaba)

yuuu commented on app/controllers/concerns/commentable.rb:
```diff
@@ -0,0 +1,15 @@
+    if comment.save
+      redirect_to url_for(commentable), notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    else
+      redirect_to url_for(commentable), status: :unprocessable_entity
```
この書き方だと、リダイレクトされてしまうので、バリデーションエラーが発生した時にフォームに入力していた内容は消えてしまうと思います。renderだと難しいでしょうか？

ItohYuto replied:
renderを使用して修正しました。

yuuu commented on app/controllers/concerns/commentable.rb:
```diff
@@ -0,0 +1,15 @@
+module Commentable
+  extend ActiveSupport::Concern
+
+  def add_comment(commentable, comment_params)
```
確かにConcernでも実現は可能ですが、今後、`update` `destroy` といったアクションも共通化する可能性があることを考えると、それぞれのCommentsControllerの親クラスとなるCommentsControllerを作って、そこに共通処理を記述した方がコードとしては見やすくなります。修正を検討いただきたいです 🙏

ItohYuto replied:
'app/contorllers/comment_controller.rb'にcreateメソッドを追加しそれを各'CommentContorller'で継承してアクションを共通化しました。
親クラスにdestroyメソッドが存在しており、それも継承してしまいますが'routes.rb'で'only: :create'としているため、子クラス側でdestroyアクションが呼ばれることはないため、問題ないと考えております。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,62 @@
+    @reports = Report.order(:id).page(params[:page])
+  end
+
+  # GET /reports/1 or /reports/1.json
```
`respond_to` を消したので `or /reports/1.json` は消しておきましょう。他のアクションも同様です

ItohYuto replied:
修正しました。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,11 @@
+
+class CommentsController < ApplicationController
+  def destroy
+    commentable = Comment.find(params[:id]).commentable
```
ここを `current_user.comments.find(params[:id])` とすれば次の行はいらなくなりますね。
`commentable` は `comment.commentable` で参照可能です。

ItohYuto replied:
修正しました。

ItohYuto committed:
指摘内容の修正

- 不要なコメントの削除(json出力用コメント削除)
- コメント追加時にバリデーションエラーが発生した場合に入力内容を保持するように変更。
- コメント作成時のアクションを共通化
- コメントのバリデーションを追加。(htmlにrequired属性を追加しているが、アプリケーション側でもバリデーションを実装。)
- コメント削除時の処理を簡略化。

ItohYuto commented:
レビューありがとうございます。
修正いたしましたので、再度レビューお願いいたします。

## ページのスクリーンショット
- 日報一覧
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/4d600c5f-f61b-438e-96de-b3926fadc30e)
- 日報詳細ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/247a10fb-2e8e-42a9-ae3c-d11aa5bd3631)
- 日報編集ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/95d2a03c-f23d-4bba-8227-f63518f85425)
- 本詳細ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/18cb1c21-8b5e-43bc-946c-b99c8a709cdf)
## rubocop
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/c6bd8587-8076-4a05-9bb5-0e6bc99ca479)

## erblint
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/2b319d60-eae8-4596-955a-805db9c7146f)

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -1,11 +1,28 @@
   def destroy
-    commentable = Comment.find(params[:id]).commentable
-    redirect_to commentable if !comment = current_user.comments.find(params[:id])
+    redirect_to comment.commentable if !comment = current_user.comments.find(params[:id])
```
`find` を使うのであれば、対象が存在しない時点で `ActiveRecord::RecordNotFound` が `raise` されて404をレスポンスするので、ifは不要ですね。

ItohYuto replied:
修正しました。

ItohYuto committed:
不要なif文の削除

ItohYuto commented:
修正いたしましたので再度レビューお願いいたします。
今回の修正による動作の違いがwebのスクリーンショットだとわかりにくいため、ブラウザのコンソールのスクリーンショットを添付いたします。
viewに関しては変更していないため、erblintは実行しておりません。
## 存在しないidを指定してdeleteアクションを実行したときの動き

![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/90d98138-c701-4a71-a156-898e5130d56d)
## rubocop
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/b35201f9-623a-4043-9a68-d01578737a4b)
